### PR TITLE
Finalize previous MachineAutoscaler targets on update

### PR DESCRIPTION
**Pull request by bison**
This records the most recently observed target of a MachineAutoscaler in the status.  This is used to detect an update to the target and remove autoscaling configuration from the previously targeted resource before applying configuration to the new target.

### CODE REVIEW

**Comment by bison** [comment]
/hold

This is a WIP. Needs testing.

File: pkg/controller/machineautoscaler/machinetarget.go [hunk 1]
```diff
@@ -186,3 +186,31 @@ func (mt *MachineTarget) GetOwner() (types.NamespacedName, error) {
 
 	return nn, nil
 }
+
+// RemoveAnnotations removes the annotations with the given keys from the
+// target.  It returns a bool indicating whether the annotations were actually
```
Comment (frobware) [inline pkg/controller/machineautoscaler/machinetarget.go]: Just curious about how the caller distinguishes between found/not found when multiple annotations are passed but only some are deleted.

Comment (bison) [reply pkg/controller/machineautoscaler/machinetarget.go]: I'm not really sure this is a great API... We can change it at some point. In any case, this is really just meant to signal that the annotations changed in some way -- nothing more specific than that. It makes it easier to know if you need to update the resource with the API or not.

File: pkg/controller/machineautoscaler/machineautoscaler_controller.go [hunk 2]
```diff
@@ -43,11 +44,45 @@ var SupportedTargetGVKs = []schema.GroupVersionKind{
...
 
+// Config represents the configuration for a reconciler instance.
+type Config struct {
+	// The namespace for MachineAutosclaers and their targets.
+	Namespace string
+}
+
+// ObjectReference represents a CrossVersionObjectReference
+// identifying a target resource with additional helper methods.
+type ObjectReference v1alpha1.CrossVersionObjectReference
```
Comment (enxebre) [inline pkg/controller/machineautoscaler/machineautoscaler_controller.go]: Coudn't we use https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/objectreference.go?

Comment (bison) [reply pkg/controller/machineautoscaler/machineautoscaler_controller.go]: Yep. Missed that. Done.

**Pull request merged**